cp-168: delete account functionality

juanc-escobar opened the pull request:
This PR depends on #75  https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/pull/75

juanc-escobar commented:
![cp-168](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/107894952/fe7600f5-dfac-4a49-8845-9a1c37eb37dc)

juanc-escobar commented:
Hi, @Liza-Veis, @fshabanov, @v0ldemar01 could you review this draft while the profile settings get merged?

Let me know if it is posible.

Thanks

v0ldemar01 commented:
> Hi, @Liza-Veis, @fshabanov, @v0ldemar01 could you review this draft while the profile settings get merged?
> 
> Let me know if it is posible.
> 
> Thanks

Firstly, please fix all actions

OlesiaTr commented on backend/src/packages/users/libs/constants.ts/constants.ts:
```diff
@@ -0,0 +1,3 @@
+const ROWS_DELETED_SUCCESSFULLY = 0;
+
+export { ROWS_DELETED_SUCCESSFULLY };
```
The folder shouldnt be named constants.ts

OlesiaTr commented on backend/src/packages/users/user.repository.ts:
```diff
@@ -2,6 +2,7 @@ import { type Repository } from '#libs/types/types.js';
 import { UserEntity } from '#packages/users/user.entity.js';
 import { type UserModel } from '#packages/users/users.js';
 
+import { ROWS_DELETED_SUCCESSFULLY } from './libs/constants.ts/constants.js';
```
Here the file named `constants.js `

OlesiaTr commented on frontend/src/libs/components/modal/modal.tsx:
```diff
@@ -0,0 +1,90 @@
+import React from 'react';
+
+import { IconColor } from '#libs/enums/enums.js';
+import { getValidClassNames } from '#libs/helpers/get-valid-class-names.js';
```
```suggestion
import { getValidClassNames } from '#libs/helpers/helpers.js';
```

OlesiaTr commented on frontend/src/libs/components/modal/steps/delete-account/delete-account-message.tsx:
```diff
@@ -0,0 +1,34 @@
+
+import styles from './styles.module.scss';
+
+type DeleteAccountMessageProperties = {
```
As far as I noticed, in the project we use just 
```suggestion
type Properties = {
```

OlesiaTr commented on frontend/src/libs/components/modal/steps/delete-account/delete-account.ts:
```diff
@@ -0,0 +1,3 @@
+export { DeleteAccount } from './delete-account-confirmation.js';
```
The variable and file should have the same name

OlesiaTr commented on frontend/src/libs/components/modal/steps/delete-account/delete-account.ts:
```diff
@@ -0,0 +1,3 @@
+export { DeleteAccount } from './delete-account-confirmation.js';
+export { DeleteAccountMessage } from './delete-account-message.js';
+export { DeleteAccountForm } from './delete-acount-from.js';
```
```suggestion
export { DeleteAccountForm } from './delete-acount-form.js';
```

juanc-escobar commented:
Eliminated the mock button, implemented the real component, and fixed the comments. 

![cp-168 v2](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/107894952/e18514e0-3943-4780-b4d2-e6e2fb333511)

OlesiaTr commented on frontend/src/libs/components/user-profile/user-profile.tsx:
```diff
@@ -33,6 +39,25 @@ const UserProfile: React.FC = () => {
+  const steps = [
+    {
+      component: <DeleteAccountMessage />,
+      title: 'We are Sad that you are Leaving',
```
```suggestion
      title: 'We are sad that you are leaving',
```

liza-veis commented:
unlink your pr from issue

liza-veis commented on backend/src/packages/auth/auth.controller.ts:
```diff
@@ -236,6 +247,15 @@ class AuthController extends BaseController {
     };
   }
+
+  private async deleteUser(
```
```suggestion
  private async deleteAuthenticatedUser(
```

liza-veis commented on backend/src/packages/auth/auth.service.ts:
```diff
@@ -87,6 +88,31 @@ class AuthService {
     return user;
   }
+
+  public async delete(id: number): Promise<UserDeleteResponseDto> {
```
```suggestion
  public async deleteAuthenticatedUser(id: number): Promise<UserDeleteResponseDto> {
```

liza-veis commented on backend/src/packages/auth/auth.service.ts:
```diff
@@ -87,6 +88,31 @@ class AuthService {
   }
+
+  public async delete(id: number): Promise<UserDeleteResponseDto> {
+    const user = await this.userService.findById(id);
+    if (!user) {
+      throw new UsersError({
+        status: HTTPCode.NOT_FOUND,
+        message: ExceptionMessage.USER_NOT_FOUND,
+      });
+    }
```
this should be throw in user service delete method
```suggestion
```

Liza-Veis commented on backend/src/packages/auth/auth.service.ts:
```diff
@@ -87,6 +88,31 @@ class AuthService {
+      });
+    }
+
+    const deleteResult = await this.userService.delete(id);
```
```suggestion
    const deletedUser = await this.userService.delete(id);
```

liza-veis commented on backend/src/packages/auth/auth.service.ts:
```diff
@@ -87,6 +88,31 @@ class AuthService {
+    if (!deleteResult) {
+      throw new UsersError({
+        status: HTTPCode.INTERNAL_SERVER_ERROR,
+        message: 'Failed to delete user',
```
Use ExceptionMessage enum

liza-veis commented on backend/src/packages/auth/auth.service.ts:
```diff
@@ -87,6 +88,31 @@ class AuthService {
+      });
+    }
+
+    const userResponse: UserDeleteRequestDto = {
+      id: user.id,
+    };
+
+    return { user: userResponse };
```
```suggestion
    return { id: user.id };
```

dmytro-kucherenko commented on backend/src/packages/users/libs/constants/constants.ts:
```diff
@@ -0,0 +1,3 @@
+const ROWS_DELETED_SUCCESSFULLY = 0;
```
It's better to name it `NO_DELETED_ROWS` or smth like that, because it can be successful only in comparison, so the constant can't describe it

dmytro-kucherenko commented on backend/src/packages/users/user.service.ts:
```diff
@@ -95,8 +95,13 @@ class UserService implements Service {
-    return Promise.resolve(true);
+  public async delete(id: number): Promise<boolean> {
+    const user = await this.findById(id);
+    if (!user) {
+      return false;
+    }
+
+    return await this.userRepository.delete(id);
```
```suggestion
    return user && await this.userRepository.delete(id);
```
Definitely if the linter allows it

juanc-escobar replied:
I could implement it like this to avoid type problems:

return user ? await this.userRepository.delete(id) : false;

 but I believe that for the reliability of the code is better to keep the return as simple as possible. 

Implementing it like this:

return user && await this.userRepository.delete(id);

 will mean changes in types, which if required is okay, but if it can be avoided is better to keep it simple since is just small code.

dmytro-kucherenko replied:
If there's a problem with return type, you can add `Boolean(...)` to check if data exists

dmytro-kucherenko commented on frontend/src/libs/components/modal/modal.tsx:
```diff
@@ -0,0 +1,90 @@
+import styles from './styles.module.scss';
+
+const INITIAL_STEP = 0;
+const NEXT_STEP_INCREMENT = 1;
```
It's better to have separate constants file

dmytro-kucherenko commented on frontend/src/libs/components/modal/modal.tsx:
```diff
@@ -0,0 +1,90 @@
+    setCurrentStep(INITIAL_STEP);
+    if (onClose) {
+      onClose();
+    }
```
```suggestion
    onClose?.();
```

dmytro-kucherenko commented on frontend/src/libs/components/modal/modal.tsx:
```diff
@@ -0,0 +1,90 @@
+    <div
+      className={getValidClassNames(
+        styles['overlay'],
+        !isDisplayed && 'visually-hidden',
```
I slightly changed common modal with use of the `dialog` component according to the @Liza-Veis suggestion, you can check it [here](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/pull/213/files#diff-d46dbe8116df7bd9fa204c2bc382baa2d3aa61e67b81cc025b2740b41ca2d2f9)

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/delete-account-confirmation.tsx:
```diff
@@ -0,0 +1,60 @@
+      void dispatch(authActions.deleteUser(payload));
+      if (onClose) {
+        onClose();
+      }
```
The same

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/delete-account-message.tsx:
```diff
@@ -0,0 +1,31 @@
+        </button>
+        <button onClick={onNext} className={styles['button']}>
+          Continue
+        </button>
```
We should use common components, it seems that nothing interferes you from that approach

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/delete-acount-form.tsx:
```diff
@@ -0,0 +1,113 @@
+      <form className={styles['form']}>
+        <div>
+          <label>
+            <input
+              type="checkbox"
+              name="checkbox1"
+              onChange={handleCheckboxChange}
+              checked={isChecked.checkbox1}
+            />
+            I&apos;ve chose another app
```
The same, we have common checkbox component, just create another preseted style there

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/delete-acount-form.tsx:
```diff
@@ -0,0 +1,113 @@
+
+  return (
+    <div className={styles['modal-body']}>
+      <form className={styles['form']}>
```
I'm pretty sure you should validate this form using `useAppForm` and control every checkbox that way

juanc-escobar replied:
the form is currently a mockup we are not doing anything with it yet

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/styles.module.scss:
```diff
@@ -0,0 +1,69 @@
+.modal-body {
+  padding: 10px;
+  font-size: 16px;
+  font-family: Montserrat, Arial, sans-serif;
```
It's no need to set font family, it's already on body tag by default

v0ldemar01 commented on backend/src/packages/users/user.repository.ts:
```diff
@@ -106,8 +107,10 @@ class UserRepository implements Repository {
-  public delete(): ReturnType<Repository['delete']> {
-    return Promise.resolve(true);
+  public async delete(id: number): Promise<boolean> {
+    const rowsDeleted = await this.userModel.query().deleteById(id);
```
Let's add a soft deletion

juanc-escobar replied:
Hi @v0ldemar01 , I've updated all the other requested changes. However, I'm not entirely sure on what needs to be done for this particular task. Could you please provide more detailed instructions on what I need to do?

v0ldemar01 replied:
Soft deletion means that instead of deleting entity the deleted_at would be added

v0ldemar01 replied:
The same one is on miniproject

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/styles.module.scss:
```diff
@@ -0,0 +1,73 @@
+  white-space: nowrap;
+}
+
+.modal-body label {
```
According to the quality criteria we can't use tag names in styles

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/delete-acount-form.tsx:
```diff
@@ -0,0 +1,105 @@
+            disableDefaultStyles
+          />
+        </div>
+        <input
+          type="text"
```
You can use common input component

OlenaYatseniuk reviewed (approved):
It looks good to me

liza-veis commented on backend/src/db/migrations/20230915120000_add_deleted_at_to_users_table.ts:
```diff
@@ -0,0 +1,18 @@
+import { type Knex } from 'knex';
+
+const TABLE_NAME = 'users';
+const COLUMN_NAME = 'deleted_at';
```
```suggestion
const ColumnName = {
  DELETED_AT: 'deleted_at';
} as const;
```

liza-veis commented on backend/src/db/migrations/20230915120000_add_deleted_at_to_users_table.ts:
```diff
@@ -0,0 +1,18 @@
+
+async function up(knex: Knex): Promise<void> {
+  await knex.schema.table(TABLE_NAME, (table) => {
+    table.dateTime(COLUMN_NAME).nullable();
```
it's nullable by default
```suggestion
    table.dateTime(COLUMN_NAME);
```

liza-veis commented on backend/src/packages/auth/auth.controller.ts:
```diff
@@ -238,6 +249,15 @@ class AuthController extends BaseController {
     };
   }
+
+  private async deleteAuthenticatedUser(
```
we need to add this route to swagger

juanc-escobar replied:
I implemented the other comments, still working on this one.

juanc-escobar replied:
Done✅

liza-veis commented on backend/src/packages/auth/auth.service.ts:
```diff
@@ -87,6 +87,21 @@ class AuthService {
+  public async deleteAuthenticatedUser(
+    id: number,
+  ): Promise<UserDeleteResponseDto> {
+    const deleteResult = await this.userService.delete(id);
```
```suggestion
    const deletedUser = await this.userService.delete(id);
```

liza-veis commented on backend/src/packages/auth/auth.service.ts:
```diff
@@ -87,6 +87,21 @@ class AuthService {
+      });
+    }
+
+    return { id: id };
```
```suggestion
    return { id: deltedUser.id };
```

liza-veis commented on backend/src/packages/users/user.repository.ts:
```diff
@@ -106,15 +111,41 @@ class UserRepository implements Repository {
-  public delete(): ReturnType<Repository['delete']> {
-    return Promise.resolve(true);
+  public async delete(id: number): Promise<UserEntity | null> {
+    const user = await this.userModel
+
+      .query()
+      .modify('withoutPassword')
+      .withGraphJoined(UsersRelation.DETAILS)
+      .findById(id)
+      .castTo<UserCommonQueryResponse | undefined>()
+      .execute();
+
+    if (!user) {
+      return null;
+    }
```
won't patch check for user with id?
```suggestion
```

liza-veis commented on backend/src/packages/users/user.repository.ts:
```diff
@@ -106,15 +111,41 @@ class UserRepository implements Repository {
+
+    await this.userModel.query().patch({ deletedAt: new Date() }).where({ id });
+
+    user.deletedAt = new Date().toISOString();
```
you need to take it from returned result
```suggestion
```

liza-veis commented on backend/src/packages/users/user.repository.ts:
```diff
@@ -106,15 +111,41 @@ class UserRepository implements Repository {
+
+    user.deletedAt = new Date().toISOString();
+
+    return UserEntity.initialize({
+      id: user.id,
+      email: user.email,
+      createdAt: new Date(user.createdAt),
+      updatedAt: new Date(user.updatedAt),
+      fullName: user.details?.fullName ?? '',
+      isSurveyCompleted: user.details?.isSurveyCompleted ?? false,
+      deletedAt: user.deletedAt ? new Date(user.deletedAt) : null,
+    });
```
we can just return boolean instead of this
```suggestion
    return true
```

liza-veis commented on backend/src/packages/users/user.service.ts:
```diff
@@ -95,8 +95,15 @@ class UserService implements Service {
-  public delete(): ReturnType<Service['delete']> {
-    return Promise.resolve(true);
+  public async delete(id: number): Promise<boolean> {
+    const user = await this.findById(id);
+    if (!user) {
+      return false;
+    }
```
we should throw an error in case there is no user with this id

liza-veis commented on frontend/src/libs/components/modal/steps/delete-account/delete-acount-form.tsx:
```diff
@@ -0,0 +1,104 @@
+
+  return (
+    <div className={styles['modal-body']}>
+      <form className={styles['form']}>
```
all forms should be controlled using useAppForm

juanc-escobar replied:
I implemented the other comments, still working on this one.

juanc-escobar replied:
Done✅

v0ldemar01 commented on backend/src/packages/auth/auth.service.ts:
```diff
@@ -87,6 +87,21 @@ class AuthService {
+
+    if (!deletedUser) {
+      throw new UsersError({
+        status: HTTPCode.INTERNAL_SERVER_ERROR,
```
That case is suitable for 404

v0ldemar01 commented on backend/src/packages/users/libs/constants/constants.ts:
```diff
@@ -0,0 +1,3 @@
+const ZERO_ROWS_DELETED = 0;
```
Maybe rename it

v0ldemar01 commented on backend/src/packages/users/user.repository.ts:
```diff
@@ -106,15 +112,21 @@ class UserRepository implements Repository {
+  public async delete(id: number): Promise<boolean> {
+    const updatedRows = await this.userModel
+      .query()
+      .patch({ deletedAt: new Date() })
```
```suggestion
      .patch({ deletedAt: new Date().toISOString() })
```

v0ldemar01 commented on backend/src/packages/auth/auth.service.ts:
```diff
@@ -87,6 +87,21 @@ class AuthService {
     return user;
   }
+
+  public async deleteAuthenticatedUser(
```
I think auth.service is not responsible for deleting user

juanc-escobar replied:
I placed the delete account functionality in the auth controller and service because to me it seemed authentication related (create an account, delete account), since deleting an account often requires the user to be authenticated, it makes sense for the delete functionality to be a part of the AuthService as this service handles authenticated actions.

Also in a commercial app account deletion often requires additional security checks like password revalidation. Since AuthService is handling other security-related operations, it might be good to leave it there for future changes in the functionality.

That's what I thought but if it needs to be moved it can be done, Do you want me to leave it in auth controller or do you want me to create a user controller? or maybe move it elsewhere, let me know and I will work on it.

juanc-escobar replied:
@v0ldemar01

v0ldemar01 replied:
I read your arguments, but let's handle it in user controller

v0ldemar01 commented on frontend/src/libs/components/checkbox/checkbox.tsx:
```diff
@@ -2,19 +2,37 @@ import styles from './styles.module.scss';
+  onChange,
+  disableDefaultStyles,
+}) => {
+  const checkboxClassName = disableDefaultStyles ? '' : styles['checkbox'];
+  const containerClassName = disableDefaultStyles ? '' : styles['container'];
+  const labelClassName = disableDefaultStyles ? '' : styles['label'];
+  const checkboxContainerClassName = disableDefaultStyles
+    ? styles['checkbox-container']
+    : '';
```
Use getValidClassNames here

v0ldemar01 commented on frontend/src/libs/components/input/input.tsx:
```diff
@@ -20,6 +20,7 @@ type Properties<T extends FormFieldValues> = {
   maxLength?: number;
   autoComplete?: 'off' | 'on';
   isChatInput?: boolean;
+  disabled?: boolean;
```
Prefer the acceptance criteria rule here

juanc-escobar replied:
I don't think I fully understand what you mean here, can you please provide me a little more context so I can understand what is required?

juanc-escobar replied:
@v0ldemar01

v0ldemar01 replied:
I mean B19 rule of quality criteria

v0ldemar01 commented on frontend/src/libs/components/modal/steps/delete-account/delete-account-confirmation.tsx:
```diff
@@ -0,0 +1,57 @@
+  const handleDelete = useCallback(
+    (payload: UserDeleteRequestDto): void => {
+      void dispatch(authActions.deleteUser(payload));
+      onClose?.();
```
I suppose we should also handle logout for auth user after deleting

juanc-escobar replied:
```typescript
const deleteUser = createAsyncThunk<
  unknown,
  UserDeleteRequestDto,
  AsyncThunkConfig
>(`${sliceName}/delete-user`, async (deletePayload, { extra, dispatch }) => {
  const { authApi, storage } = extra;
  const { id } = await authApi.deleteUser(deletePayload);

  await storage.drop(StorageKey.TOKEN);

  dispatch(appActions.navigate(AppRoute.SIGN_IN));

  return id;
});
```

I did it here frontend\src\slices\auth\actions.ts

I removed the token from the user and redirected it to the sign-in page so the user can no longer sign in to the page once deleted, let me know if it is ok or if you need me to do this elsewhere.

juanc-escobar replied:
@v0ldemar01

v0ldemar01 commented on frontend/src/libs/components/modal/steps/delete-account/delete-acount-form.tsx:
```diff
@@ -0,0 +1,109 @@
+
+  const isNextDisabled = !Object.values(checkboxValues).some(Boolean);
+
+  const checkboxes = [
```
Move to constants

v0ldemar01 commented on frontend/src/pages/user-profile/components/user-profile-sidebar/user-profile-sidebar.tsx:
```diff
@@ -51,6 +58,25 @@ const UserProfileSidebar: React.FC<Properties> = ({
+    setModalOpen(!isModalOpen);
+  }, [isModalOpen]);
+
+  const steps = [
+    {
+      component: <DeleteAccountMessage />,
+      title: 'We are sad that you are leaving',
+    },
+    {
+      component: <DeleteAccountForm />,
+      title: 'Please tell us why',
+    },
+    {
+      component: <DeleteAccountConfirmation />,
+      title: 'Your account will be deleted',
+    },
+  ];
```
```suggestion
  const steps = [
    {
      component: DeleteAccountMessage,
      title: 'We are sad that you are leaving',
    },
    {
      component: DeleteAccountForm,
      title: 'Please tell us why',
    },
    {
      component: DeleteAccountConfirmation,
      title: 'Your account will be deleted',
    },
  ];
```

v0ldemar01 replied:
And also move out

dmytro-kucherenko commented on frontend/src/libs/components/modal/modal.tsx:
```diff
@@ -0,0 +1,93 @@
+    : null;
+
+  return (
+    <dialog
+      open={isDisplayed}
```
Check how we are handling references at now [here](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/pull/213/files#diff-d46dbe8116df7bd9fa204c2bc382baa2d3aa61e67b81cc025b2740b41ca2d2f9) as @v0ldemar01 suggested, because we have to synchronize our effortsю The same approach should be used in your modal wrapper

dmytro-kucherenko commented on backend/src/packages/users/user.service.ts:
```diff
@@ -95,8 +98,21 @@ class UserService implements Service {
+  public async delete(
+    id: number,
+  ): Promise<ReturnType<UserEntity['toObject']> | null> {
+    const deletedUser = await this.userRepository.findById(id);
+
+    if (!deletedUser) {
+      throw new UsersError({
+        status: HTTPCode.NOT_FOUND,
+        message: ExceptionMessage.USER_NOT_FOUND,
+      });
+    }
+
+    const isDeleted = await this.userRepository.delete(id);
+
+    return isDeleted ? deletedUser.toObject() : null;
```
Can you also patch `deletedAt` date in repo only for `whereNull('deletedAt')`? Then you needn't check `deletedUser` and can throw error right after `isDeleted === false` and return non nullable object
Anyway you're throwing error in auth service, so there's no matter to divide this functionality

![image](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/92298163/37be8a6e-0a0b-49f8-802a-f446b83aa29d)

dmytro-kucherenko commented on frontend/src/libs/components/card/card.tsx:
```diff
@@ -44,7 +48,12 @@ const Card: React.FC<Properties> = ({
             <div className={styles['icon-background']}>
-              <Icon name={iconName as IconName} color={iconColor} />
+              <Icon
+                name={iconName as IconName}
+                color={iconColor as ValueOf<typeof IconColor>}
+                width={iconWidth as number}
+                height={iconHeight as number}
```
I setted `Icon` component props as that can be `undefined`, while you made these assertions.
@v0ldemar01, @fshabanov What will be the better approach to handle this issue?

dmytro-kucherenko commented on frontend/src/libs/components/checkbox/checkbox.tsx:
```diff
@@ -2,19 +2,37 @@ import styles from './styles.module.scss';
-  onChange: () => void;
+  checked?: boolean;
+  name?: string;
+  onChange: (event: React.ChangeEvent<HTMLInputElement>) => void;
```
You don't need event as I see from your usage

![image](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/92298163/a1e5d5fd-b73b-428b-8a56-3c674e5adc8e)

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/libs/types/delete-account-form-payload.ts:
```diff
@@ -0,0 +1,11 @@
+type DeleteAccountFormPayload = {
```
I think it should be stored in shared because mobile side also has to log out some way)

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/libs/types/delete-account-form-payload.ts:
```diff
@@ -0,0 +1,11 @@
+type DeleteAccountFormPayload = {
+  describeYourSituation: string;
+  checkboxes: {
+    checkbox1: boolean;
+    checkbox2: boolean;
+    checkbox3: boolean;
+    checkbox4: boolean;
```
This approach is difficult to extend, because for any new reason option we need to create new checkbox prop in dto. Please check how mupltiple checkboxes were handled in survey page [here](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/pull/89/files#diff-92edabe5fdb8ef427ad887499a7713ef8d3a1a13638ab7bbf1780af60806318b) on the line 41, they all have the same name and are validated as [array of strings](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/pull/89/files#diff-c89514f464530a1aae8beb2bd048c37ab23832896c9f2d58e85fafc55f6d958d)

dmytro-kucherenko commented on frontend/src/libs/components/checkbox/checkbox.tsx:
```diff
@@ -1,20 +1,46 @@
+    disableDefaultStyles ? null : styles['checkbox'],
+  );
+  const containerClassName = getValidClassNames(
+    disableDefaultStyles ? null : styles['container'],
```
```suggestion
    !disableDefaultStyles ?? styles['container'],
```
The same below

dmytro-kucherenko commented on frontend/src/libs/components/checkbox/checkbox.tsx:
```diff
@@ -1,20 +1,46 @@
+  checked,
+  name,
+  onChange,
+  disableDefaultStyles,
```
```suggestion
  isDefaultStylesDisabled,
```
Lets follow quality criteria

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/delete-account-message.tsx:
```diff
@@ -0,0 +1,29 @@
+      </ul>
+      <div className={styles['footer']}>
+        {onClose && <Button label="Cancel" style="primary" onClick={onClose} />}
+        {onNext && <Button label="Continue" style="primary" onClick={onNext} />}
```
May we create separate boolean like `hasNextStep`?

juanc-escobar replied:
I think is not necessary, steps are managed by an array and the logic for advancing steps is handled by the goToNextStep function.

v0ldemar01 replied:
Please follow @Dmytro-Kucherenko comment

juanc-escobar replied:
at first, I didn't understand correctly what you meant by the add the hasNextStep @Dmytro-Kucherenko , now i get it, thanks.

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/delete-acount-form.tsx:
```diff
@@ -0,0 +1,109 @@
+        <Input
+          control={control}
+          errors={errors}
+          name="describeYourSituation"
```
```suggestion
          name="description"
```

dmytro-kucherenko commented on frontend/src/pages/user-profile/components/user-profile-sidebar/user-profile-sidebar.tsx:
```diff
@@ -51,6 +66,46 @@ const UserProfileSidebar: React.FC<Properties> = ({
   }, [dispatch]);
 
+  const handleCloseModal = useCallback((): void => {
+    setModalOpen(false);
```
I think there's boolean that describes current dialog behaviour, as I remember `ref.current.open`. Try to find out and implement it, then you don't have to use `onClose` prop and state

v0ldemar01 replied:
@Dmytro-Kucherenko is right. Please check our latest flow with modal

dmytro-kucherenko commented on frontend/src/libs/components/checkbox/checkbox.tsx:
```diff
@@ -1,20 +1,46 @@
-        className={styles['checkbox']}
+        className={checkboxClassName}
         type="checkbox"
+        name={name}
```
```suggestion
```
You don't use control and just overwrite field onChange, so you shouldn't pass name at all

dmytro-kucherenko commented on frontend/src/libs/components/modal/libs/constants/constants.ts:
```diff
@@ -0,0 +1,48 @@
+  checkboxes: [],
+};
+
+const CHECKBOX_OPTIONS = [
+  { name: 'checkbox1', label: 'I have chosen another app' },
+  { name: 'checkbox2', label: 'I have reached my goal' },
+  { name: 'checkbox3', label: 'I can not afford the current pricing' },
+  { name: 'checkbox4', label: 'Other' },
+] as const;
```
The same about names here, you can leave only labels array

danilittaa commented on frontend/src/libs/components/checkbox/checkbox.tsx:
```diff
@@ -1,20 +1,46 @@
 
 type Properties = {
   label: string;
+  checked?: boolean;
```
```suggestion
  isChecked?: boolean;
```

danilittaa commented on frontend/src/libs/components/modal/steps/delete-account/delete-account-message.tsx:
```diff
@@ -0,0 +1,29 @@
+  onClose?: () => void;
+};
+
+const DeleteAccountMessage: FC<Properties> = ({ onNext, onClose }) => {
```
```suggestion
const DeleteAccountMessage: React.FC<Properties> = ({ onNext, onClose }) => {
```

v0ldemar01 commented on backend/src/packages/users/libs/constants/constants.ts:
```diff
@@ -0,0 +1,3 @@
+const NO_UPDATES = 0;
```
Please rename to more specific one)

v0ldemar01 commented on backend/src/packages/users/user.controller.ts:
```diff
@@ -0,0 +1,105 @@
+   *           example:
+   *            message: "User with these credentials was not found."
+   *            errorType: "USERS"
+   *        500:
```
It is useless block

v0ldemar01 commented on backend/src/packages/users/user.repository.ts:
```diff
@@ -106,15 +112,21 @@ class UserRepository implements Repository {
-  public delete(): ReturnType<Repository['delete']> {
-    return Promise.resolve(true);
+  public async delete(id: number): Promise<boolean> {
+    const updatedRows = await this.userModel
```
It can be in plural form

juanc-escobar replied:
@v0ldemar01  do you mean the name of the method delete() should I change it? or do you mean something else?

juanc-escobar replied:
@v0ldemar01  I did all the changes and updated the PR, once you provide me a little more context I can fix this too if needed.

v0ldemar01 replied:
Sorry, a little bit confusing, pass in single form, bcos only one user exists by id

v0ldemar01 commented on backend/src/packages/users/user.service.ts:
```diff
@@ -95,8 +98,26 @@ class UserService implements Service {
+  public async delete(id: number): Promise<boolean> {
+    const deletedUser = await this.userRepository.findById(id);
+
+    if (deletedUser === null) {
```
Move this block for !isDeleted case

v0ldemar01 commented on backend/src/packages/users/user.service.ts:
```diff
@@ -95,8 +98,26 @@ class UserService implements Service {
+
+    if (!isDeleted) {
+      throw new UsersError({
+        status: HTTPCode.INTERNAL_SERVER_ERROR,
```
It's useless

v0ldemar01 commented on frontend/src/libs/components/checkbox/checkbox.tsx:
```diff
@@ -1,20 +1,43 @@
+  onChange,
+  isDefaultStylesDisabled,
+}) => {
+  const checkboxClassName = getValidClassNames(
+    isDefaultStylesDisabled ? null : styles['checkbox'],
+  );
+  const containerClassName = getValidClassNames(
+    isDefaultStylesDisabled ? null : styles['container'],
+  );
+  const labelClassName = getValidClassNames(
+    isDefaultStylesDisabled ? null : styles['label'],
+  );
+  const checkboxContainerClassName = getValidClassNames(
+    isDefaultStylesDisabled ? styles['checkbox-container'] : null,
+  );
```
```suggestion
  const checkboxClassName = getValidClassNames(
    !isDefaultStylesDisabled && styles['checkbox'],
  );
  const containerClassName = getValidClassNames(
    !isDefaultStylesDisabled && styles['container'],
  );
  const labelClassName = getValidClassNames(
    !isDefaultStylesDisabled && styles['label'],
  );
  const checkboxContainerClassName = getValidClassNames(
    sDefaultStylesDisabled && styles['checkbox-container'],
  );
```

v0ldemar01 commented on frontend/src/libs/components/modal/steps/delete-account/delete-account-confirmation.tsx:
```diff
@@ -0,0 +1,55 @@
+        Thank you for your feedback and for being with us
+      </p>
+      <p className={styles['text']}>
+        are you sure you want to delete your account?
```
```suggestion
        Are you sure you want to delete your account?
```

v0ldemar01 commented on frontend/src/libs/components/modal/steps/delete-account/delete-acount-form.tsx:
```diff
@@ -0,0 +1,106 @@
+    [checkedBoxes, onCheckboxChange],
+  );
+
+  const isInputDisabled = !checkedBoxes.includes('checkbox4');
```
checkbox4 => please add more specific name

v0ldemar01 commented on frontend/src/libs/components/modal/steps/delete-account/delete-acount-form.tsx:
```diff
@@ -0,0 +1,106 @@
+            <div key={checkbox.label}>
+              <Checkbox
+                label={checkbox.label}
+                isChecked={checkedBoxes.includes(checkbox.label) || false}
```
```suggestion
                isChecked={checkedBoxes.includes(checkbox.label)}
```

v0ldemar01 commented on frontend/src/libs/components/modal/steps/delete-account/delete-acount-form.tsx:
```diff
@@ -0,0 +1,106 @@
+          isDisabled={isInputDisabled}
+        />
+        <div className={styles['footer']}>
+          {onClose && (
```
Add boolean for it

v0ldemar01 commented on frontend/src/pages/user-profile/components/user-profile-sidebar/user-profile-sidebar.tsx:
```diff
@@ -51,6 +66,46 @@ const UserProfileSidebar: React.FC<Properties> = ({
+    return <StepComponent onClose={handleCloseModal} onNext={goToNextStep} />;
+  };
+
+  useEffect(() => {
+    if (modalReference.current) {
+      if (isModalOpen) {
+        modalReference.current.showModal();
+      } else {
+        modalReference.current.close();
+      }
+    }
+  }, [isModalOpen, modalReference]);
```
```suggestion
  useEffect(() => {
     if (isModalOpen) {
        modalReference.current?.showModal();
      } else {
        modalReference.current?.close();
      }
  }, [isModalOpen, modalReference]);
```

v0ldemar01 commented on frontend/src/pages/user-profile/components/user-profile-sidebar/user-profile-sidebar.tsx:
```diff
@@ -51,6 +66,46 @@ const UserProfileSidebar: React.FC<Properties> = ({
+    }
+  }, [isModalOpen, modalReference]);
+
+  const toggleModal = useCallback(() => {
```
Please follow acceptance criteria

v0ldemar01 commented on frontend/src/pages/user-profile/components/user-profile-sidebar/user-profile-sidebar.tsx:
```diff
@@ -92,6 +147,23 @@ const UserProfileSidebar: React.FC<Properties> = ({
+            iconWidth={24}
+            iconHeight={24}
+          />
+          {currentStep && (
```
Add boolean for it

v0ldemar01 commented on frontend/tsconfig.json:
```diff
@@ -1,6 +1,9 @@
-  "include": ["src"],
+  "include": [
+    "src",
+    "../shared/src/packages/users/libs/types/delete-account-form-payload.type.ts"
```
What is the purpose for it?

juanc-escobar replied:
@v0ldemar01 I don't recall manually adding that particular configuration. While I did unmount the Docker container and delete the database when switching branches, I don't remember modifying the tsconfig.json file. I will delete it.

fshabanov commented on backend/src/packages/users/user.controller.ts:
```diff
@@ -0,0 +1,96 @@
+  type APIHandlerOptions,
+  type APIHandlerResponse,
+} from '#libs/packages/controller/controller.js';
+import { BaseController } from '#libs/packages/controller/controller.js';
```
Combine imports

OlesiaTr commented on frontend/src/libs/components/checkbox/checkbox.tsx:
```diff
@@ -1,34 +1,39 @@
-  disableDefaultStyles?: boolean;
+  isChecked?: boolean;
+  onChange: () => void;
+  isDefaultStylesDisabled?: boolean;
```
```suggestion
  isDefaultStyleDisabled?: boolean;
```

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/delete-account-message.tsx:
```diff
@@ -0,0 +1,34 @@
+};
+
+const DeleteAccountMessage: React.FC<Properties> = ({ onNext, onClose }) => {
+  const hasNextStep = onNext && typeof onNext === 'function';
+  const hasCloseButton = onClose && typeof onClose === 'function';
```
```suggestion
  const hasNextStep = Boolean(onNext);
  const hasCloseButton = Boolean(onClose);
```

juanc-escobar replied:
I have tried to do that previously but always gives me this error:

![image](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/107894952/3fb31440-cfd7-4e90-9c44-6f0edf9bd616)

I can solve it by adding the option of undefined to the onClick property on the button:

![image](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/107894952/8e09bebd-4098-4b56-9276-2ab06773982e)

but I am not sure if adding undefined to the onClick will affect other functionality since it is a component that is used frequently. I will make the change but I had chosen before to specify the type to avoid changing the original button component and avoid any possible bug.

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/delete-acount-form.tsx:
```diff
@@ -0,0 +1,109 @@
+  const isInputDisabled = !checkedBoxes.includes(OTHER_OPTION_LABEL);
+  const isNextDisabled = checkedBoxes.length === NO_CHECKED_BOXES;
+  const hasNextStep = onNext && typeof onNext === 'function';
+  const hasCloseButton = onClose && typeof onClose === 'function';
```
The same as above

dmytro-kucherenko commented on frontend/src/libs/components/modal/steps/delete-account/styles.module.scss:
```diff
@@ -0,0 +1,65 @@
+  margin-top: 45px;
+}
+
+.input {
+  width: 100%;
+  height: 50px;
+  font-size: 14px;
+  font-family: Montserrat, Arial, sans-serif;
+  line-height: 1.25;
+  border-radius: 8px;
+}
+
+.button {
+  display: inline-block;
+  width: 100%;
+  padding: 15px 0;
+  color: var(--white);
+  font-weight: 600;
+  font-size: 16px;
+  font-family: Montserrat, Arial, sans-serif;
+  line-height: 1.25;
+  white-space: nowrap;
+  background-color: var(--gray-400);
+  border: none;
+  border-radius: 8px;
+  cursor: pointer;
+}
+
+.button:hover {
+  background-color: var(--gray-400);
+}
+
+.button:disabled {
+  color: var(--gray-200);
+  background-color: var(--gray-300);
+  cursor: not-allowed;
+}
```
```suggestion
```

You don't use them

dmytro-kucherenko commented on frontend/src/slices/users/actions.ts:
```diff
@@ -0,0 +1,26 @@
+import { name as sliceName } from './users.slice.js';
+
+const deleteUser = createAsyncThunk<
+  unknown,
```
```suggestion
  number,
```
You return id. If you don't need it, just make it void

v0ldemar01 commented on backend/src/packages/users/user.service.ts:
```diff
@@ -95,8 +98,17 @@ class UserService implements Service {
-  public delete(): ReturnType<Service['delete']> {
-    return Promise.resolve(true);
+  public async delete(id: number): Promise<boolean> {
+    const deletedUser = await this.userRepository.findById(id);
```
```suggestion
    const userToDelete = await this.userRepository.findById(id);
```

v0ldemar01 commented on backend/src/libs/types/repository.type.ts:
```diff
@@ -3,7 +3,7 @@ type Repository<T = unknown> = {
   create(payload: unknown): Promise<T>;
   update(payload: unknown): Promise<T>;
-  delete(payload: unknown): Promise<number>;
+  delete(payload: unknown): Promise<number | boolean>;
```
```suggestion
  delete(payload: unknown): Promise<number>;
```

v0ldemar01 commented on backend/src/packages/users/user.repository.ts:
```diff
@@ -133,18 +141,21 @@ class UserRepository implements Repository {
-  public delete(): ReturnType<Repository['delete']> {
-    //TODO
-    const deletedCount = 0;
+  public async delete(id: number): Promise<boolean> {
+    const updatedRow = await this.userModel
+      .query()
+      .patch({ deletedAt: new Date().toISOString() })
+      .where({ id, deletedAt: null });
```
```suggestion
  public delete(id: number): Promise<boolean> {
    return this.userModel
      .query()
      .patch({ deletedAt: new Date().toISOString() })
      .where({ id, deletedAt: null })
      .execute();
```

v0ldemar01 commented on backend/src/packages/users/libs/constants/constants.ts:
```diff
@@ -0,0 +1,3 @@
+const ZERO_UPDATED_ROW_SOFT_DELETED = 0;
```
Get rid of it

v0ldemar01 commented on backend/src/packages/users/user.service.ts:
```diff
@@ -104,8 +107,17 @@ class UserService implements Service {
+  public async delete(id: number): Promise<boolean> {
+    const userToDelete = await this.userRepository.findById(id);
+
+    if (userToDelete === null) {
```
```suggestion
    if (!userToDelete) {
```

v0ldemar01 commented on frontend/src/pages/user-profile/components/user-profile-sidebar/user-profile-sidebar.tsx:
```diff
@@ -43,6 +52,12 @@ const UserProfileSidebar: React.FC<Properties> = ({
+  const dialogReference = useRef<HTMLDialogElement>(null);
+  const currentStep = STEPS[currentStepIndex];
+  const hasCurrentStep =
+    currentStep && typeof currentStep.component === 'function';
```
```suggestion
    currentStep;
```

v0ldemar01 commented on frontend/src/pages/user-profile/components/user-profile-sidebar/user-profile-sidebar.tsx:
```diff
@@ -58,6 +73,34 @@ const UserProfileSidebar: React.FC<Properties> = ({
+  }, []);
+
+  const getCurrentStep = (): ReactElement | null => {
+    if (!currentStep) {
+      return null;
+    }
```
```suggestion
```

juanc-escobar replied:
Hi, @v0ldemar01 for this one I need to check that the current step exists and is not undefined otherwise, the typescript throws an error:

![image](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/107894952/8d817eec-0d1f-4d6f-a1ed-65f42b4332d5)

I updated the function like this to avoid the if statement but I am not sure it complies with what is requested, if it is not can you give me a suggestion to avoid this typescript error? 

![image](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/107894952/2cf7c2a5-5638-4b7a-86f7-6f8a0c98d439)

v0ldemar01 reviewed (changes requested):
Last small improvements

juanc-escobar commented:
@v0ldemar01  also got this error from Continuous Integration Postman Run Tests Development / Run latest database migration on environment dump (pull_request) Failing after 21s
:

![image](https://github.com/BinaryStudioAcademy/bsa-2023-calmpal/assets/107894952/2be09042-5bd8-4202-8ff8-9a18d22593a6)

the error is related to chat Validation of schema but I have not touched that.

v0ldemar01 commented on frontend/src/pages/user-profile/components/user-profile-sidebar/user-profile-sidebar.tsx:
```diff
@@ -43,6 +52,11 @@ const UserProfileSidebar: React.FC<Properties> = ({
+
+  const dialogReference = useRef<HTMLDialogElement>(null);
+  const currentStep = STEPS[currentStepIndex];
+  const hasCurrentStep = currentStep;
```
```suggestion
```

v0ldemar01 commented on frontend/src/pages/user-profile/components/user-profile-sidebar/user-profile-sidebar.tsx:
```diff
@@ -105,6 +145,19 @@ const UserProfileSidebar: React.FC<Properties> = ({
+            iconWidth={24}
+            iconHeight={24}
+          />
+          {hasCurrentStep && (
```
```suggestion
          {Boolean(currentStep)&& (
```

v0ldemar01 commented on frontend/src/pages/user-profile/components/user-profile-sidebar/user-profile-sidebar.tsx:
```diff
@@ -58,6 +72,32 @@ const UserProfileSidebar: React.FC<Properties> = ({
+  }, []);
+
+  const getCurrentStep = (): ReactElement | null => {
+    const StepComponent = currentStep?.component;
```
You are checking Boolean(currentStep) before calling this function. Use type assertion for this case

fshabanov commented on frontend/src/libs/components/modal/libs/constants/constants.ts:
```diff
@@ -0,0 +1,52 @@
+import { type DeleteAccountFormPayload } from 'shared/build/packages/users/users.js';
```
Import from #packages/users/users.js

fshabanov commented on shared/src/packages/users/libs/enums/users-api-path.enum.ts:
```diff
@@ -1,5 +1,6 @@
 const UsersApiPath = {
   ROOT: '/',
+  DELETE_USER: '/delete-user/:id',
```
```suggestion
  $ID: '/:id',
```
I think it would be better to send DELETE request to /:id endpoint

dmytro-kucherenko commented on frontend/src/packages/users/libs/api/users-api.ts:
```diff
@@ -0,0 +1,43 @@
+  public async deleteUser(
+    payload: UserDeleteRequestDto,
+  ): Promise<UserDeleteResponseDto> {
+    const updatedPath = UsersApiPath.DELETE_USER.replace(
+      ':id',
+      payload.id.toString(),
+    );
+    const deleteUserUrl = this.getFullEndpoint(updatedPath, {});
+
+    const response = await this.load(deleteUserUrl, {
```
```suggestion
    const response = await this.load(
      this.getFullEndpoint(UsersApiPath.$ID, { id: payload.id.toString() }),
      {
```

fshabanov commented on frontend/src/libs/components/modal/libs/constants/constants.ts:
```diff
@@ -0,0 +1,51 @@
+  DeleteAccountConfirmation,
+  DeleteAccountForm,
+  DeleteAccountMessage,
+} from '#libs/components/modal/steps/delete-account/delete-account.js';
```
```suggestion
} from '#libs/components/components.js';
```

liza-veis commented on backend/src/packages/users/user.model.ts:
```diff
@@ -27,6 +27,8 @@ class UserModel extends AbstractModel {
 
   public passwordSalt!: string;
 
+  public deletedAt?: Date | string;
```
```suggestion
  public deletedAt!: Date | null;
```

liza-veis commented on frontend/src/libs/components/card/card.tsx:
```diff
@@ -12,6 +12,8 @@ type Properties = {
   isActive?: boolean;
   iconName?: IconName;
   iconColor?: ValueOf<typeof IconColor>;
+  iconWidth?: number;
+  iconHeight?: number;
```
I believe we should have the same icon width and height for all card components, no need to pass them via props
```suggestion
```

liza-veis commented on frontend/src/libs/components/checkbox/checkbox.tsx:
```diff
@@ -1,20 +1,43 @@
+  label,
+  isChecked,
+  onChange,
+  isDefaultStyleDisabled,
```
bad decision, why can't we use default checkbox styles? we should pass other properties like color, isRounded etc. and in case it's something with different layout than use style property as for buttons
```suggestion
```

juanc-escobar replied:
hi, @Liza-Veis Please check these changes, I will try to fix the other comments as soon as I can.

liza-veis commented on frontend/src/libs/components/modal/steps/delete-account/delete-account-confirmation.tsx:
```diff
@@ -0,0 +1,55 @@
+import { Button } from '#libs/components/components.js';
```
should be in pages/profile-settings/components/delete-account-modal/delete-account-modal.tsx 

1. we need to export only one component
2. all these logic with steps should be moved there, it's a common modal component, it shouldn't know about other features

juanc-escobar replied:
hi, @Liza-Veis Please check these changes, I will try to fix the other comments as soon as I can.

liza-veis commented on shared/src/packages/users/libs/types/user-delete-response-dto.type.ts:
```diff
@@ -0,0 +1,5 @@
+type UserDeleteResponseDto = {
```
check responses in other places with delete functionality in our project, as I remember we return `isDeleted: boolean` instead of id

liza-veis commented on frontend/src/libs/types/checkbox-style.type.ts:
```diff
@@ -0,0 +1,3 @@
+type CheckboxStyle = 'primary' | 'secondary';
```
should be in checkbox component
